File: src/main/java/com/indico/request/GraphQLRequest.java
package com.indico.request;

import com.apollographql.apollo.exception.ApolloException;
import com.indico.IndicoClient;
import com.indico.RestRequest;
import com.indico.JSON;
import java.io.IOException;
import okhttp3.MediaType;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;
import org.json.JSONArray;
import org.json.JSONObject;

public class GraphQLRequest implements RestRequest<JSONObject> {

    private final IndicoClient client;
    private String query;
    private String operationName;
    private JSONObject variables;

    public GraphQLRequest(IndicoClient client) {
        this.client = client;
    }

    public GraphQLRequest query(String query) {
        this.query = query;
        return this;
    }

    public GraphQLRequest operationName(String operationName) {
        this.operationName = operationName;
        return this;
    }

    public GraphQLRequest variables(JSONObject variables) {
        this.variables = variables;
        return this;
    }

    @Override
    public JSONObject call() throws IOException {
        JSONObject json = new JSONObject();
        json.put("query", this.query);
        if (this.operationName != null) {
            json.put("operationName", this.operationName);
        }
        if (this.variables != null) {
            json.put("variables", this.variables);
        }

        String uploadUrl = this.client.config.getAppBaseUrl() + "/graph/api/graphql";
        MediaType JSON = MediaType.parse("application/json; charset=utf-8");
        RequestBody requestBody = RequestBody.create(JSON, json.toString());
        Request request = new Request.Builder()
                .url(uploadUrl)
                .post(requestBody)
                .build();

        Response result = this.client.okHttpClient.newCall(request).execute();

        String stringBody = result.body().string();
        JSONObject response = new JSON(stringBody).asJSONObject();
        result.body().close();

        if (!response.isNull("errors")) {
            JSONArray errors = response.getJSONArray("errors");
            throw new ApolloException(errors.toString());
        }
        JSONObject data = response.getJSONObject("data");
        result.close();
        return data;
    }
}

File: src/main/java/com/indico/request/GraphQLRequest.kt
package com.indico.request

import com.expediagroup.graphql.client.types.GraphQLClientRequest
import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.ObjectMapper
import com.indico.IndicoKtorClient
import com.indico.exceptions.IndicoQueryException
import com.indico.query.Query
import org.json.JSONObject
import kotlin.reflect.KClass


class RawGraphQLRequest(
    override val query: String,
    override val operationName: String,
    override val variables: JsonNode,
) : GraphQLClientRequest<JsonNode>  {
    public override fun responseType(): KClass<JsonNode> =
        JsonNode::class
}


class GraphQLRequest(
    private val indicoClient: IndicoKtorClient,
    private val query: String,
    private val operationName: String,
    private val variables: JsonNode,
) : Query<JSONObject?, JsonNode>() {

    override fun query(): JSONObject? {
         try {
            val call = RawGraphQLRequest(
                query,
                operationName,
                variables,
            )
            val response = this.indicoClient.execute(call)
            handleErrors(response)
            val objectMapper = ObjectMapper()
            val responseAsString = objectMapper.writeValueAsString(response.data)
            return JSONObject(responseAsString)
        } catch (ex: RuntimeException) {
            throw IndicoQueryException("Call to get the submission failed", ex)
        }
    }

    override fun refresh(obj: JSONObject?): JSONObject? {
        return obj
    }
}